### ISSUE TITLE
Add Webpack as an Option

### ISSUE THREAD

**Issue by pydanny** [pydanny/cookiecutter-django#538]
I've been interested in it as a way to maintain resources. And this is an AWESOME approach: https://github.com/hzdg/cookiecutter-webpack

Integrating it into Cookiecutter Django would be easy: https://github.com/hzdg/cookiecutter-webpack#integrating-into-existing-projects

**Comment by yunti** [comment]
That is very very nice, good find.

**Comment by pypetey** [comment]
+1 for that :)

**Comment by clint74** [comment]
I'm trying to understand Webpack but the documentation is pretty bad.

**Comment by pydanny** [comment]
:bell: @jambonrose

**Comment by audreyr** [comment]
I think we have to be careful with how webpack is configured if we do this. I'd prefer if Cookiecutter Django wasn't too specific to single-page apps. I lean on the side of not adding webpack unless it's added as an option and someone can maintain the feature.

**Comment by mjsisley** [comment]
I have a webpack config file for a Cookecutter Django project that essentially accomplishes most of grunt is doing for the project and adds transpiling of javascript, so it is easy to plug React in (or es6, coffeescript, etc...). It isn't specific to single-page apps. 

Here is a gist of what I have so far: [Webpack Config for Cookiecutter Django](https://gist.github.com/mjsisley/497ad917cd14ada63111e5d95951b892)

This may serve as a starting point that isn't single page app focused and works fairly similarly to how grunt is working in the project right now. The ExtractTextPlugin still needs to be configured properly to completely replace grunt (pulling css into a separate file). A lot more can be done with webpack than what is in the gist currently (ie, managing resources by chunking, [replace Django staticfile](https://github.com/owais/django-webpack-loader)).

I would like to see webpack in the project as an option (I think it makes working with React and other javascript dialects easier and allows for easier management of resources if needed) and would be happy do my best to help maintain the config file for the project if it was brought in.

If the consensus is to bring webpack in, let me know if I can help.

**Comment by jambonrose** [comment]
tldr: I agree with @audreyr 

I get the impression that Webpack is centrally used for bundling JS and assets in single page web apps (even though it could be used for bundling smaller scripts). Including it leads down a bit of a rabbit hole. If Webpack is included, a number of other pipelining decisions will likely need to be made to help/ease deployment of a single page app (such as whether to include the Django Rest Framework if Webpack is included). Ultimately, the Webpack config itself will need options, such as whether to use React, or another library, and whether to assume a particular structure for this projects. Finally, given how problematic Webpack's documentation is, I'd assume you would get quite a few requests for extra options and configurations.

Don't get me wrong: having an easy Webpack config available would be great. I just think it's going to be a lot of work, and it will position this project in a particular way.

**Comment by mjsisley** [comment]
tldr: Are you guys sure? I will submit a PR with a simple config if you guys decide you want it.

As someone who started with a Cookiecutter Django project and brought webpack in so that I could transpile React, es6, etc.., I think having a base webpack config file to start with would have been really helpful.

I worked out the issues I had with the CSS extraction to a separate path, so I could submit a pull request that provides webpack as an option with same functionality as Grunt (plus javascript transpiling). Then it would serve as an option to fulfill that role (and also handle different dialects of javascript) so that #466 could be supported and other features could be added later (but don't necessarily need to).  

As an aside, it seems you are moving towards dropping #466 and #538? Should there be a separate Cookiecutter Django - Single Page App Edition or something, or does something like this exist?

**Comment by jayfk** [comment]
> As someone who started with a Cookiecutter Django project and brought webpack in so that I could transpile React, es6, etc.., I think having a base webpack config file to start with would have been really helpful.

The main problem I see is that someone has to maintain it. There will be new releases, bugs and questions and we have to take that into account, too.

Not that I'm totally against it, we just have to keep that in mind :)

**Comment by mjsisley** [comment]
That makes complete sense @jayfk and I understand the thought process. It probably makes sense for a Django cookiecutter project to only support one frontend build tool. 

I just like webpack because of its upside and was convinced it was the best of the alternatives by [instagram's usage](https://www.youtube.com/watch?v=VkTCL6Nqm6Y) awhile ago when I wanted to add React to a Cookiecutter Django project.

I've been using it for a few months, so should have submitted a PR awhile ago. I thought someone more experienced would do it at some point. I was hoping for a more standardized approach to using React and Webpack with Django would emerge through this project.

Let me know if you do end up deciding that you want a PR on this. I would be happy to help maintain future releases, bugs, and questions (though I understand at my experience level, that may not be worth enough). If Gulp is the winner, I'll drop it now :)

**Comment by goldhand** [comment]
I found this thread this weekend and created the pull request above. I've built a couple cookiecutter projects for [HZDG](https://github.com/hzdg/), one being the [cookiecutter-webpack](https://github.com/hzdg/cookiecutter-webpack) project that I install into this project with the post hooks. It's based on our own cookiecutter-django (still private) that installs webpack in the same manner. BTW,  didn't want to jump ahead of you @mjsisley but I wanted to share how I've integrated webpack into other cookiecutter projects and get the ball rolling a little.

There are a few things I'm working on to make this a better PR but would appreciate any feedback or suggestions.

I also wanted to address the concern of @audreyr and a few others regarding webpack leaning towards spa's. It's actually very optimized for multi-page apps too. You can `"bundle"` (webpack term) all your static assets into separate modules for each page.  When webpack runs builds these assets it looks for entry points that you specify, then once it finds the entry point it loads all that entry points dependencies into the bundle, when it finds a dependency, it loads that and starts searching for it's dependencies and so on recursively. You can even include image files and style sheets as dependencies too.  The current `cookiecutter-webpack` configuration is even further optimized to load all shared assets between bundles in a single `vendor` bundle.


```javascript

  // create entry points for main js, products and users apps
  entry: {
    main: './assets/js/index',
    // apps
    'app-products': './assets/js/productApp',
    'app-users': './assets/js/userApp',
    ...
   }
  
  // all bundles will be loaded into ./assets/bundles/[name]-[hash].js
  // e.g. ./assets/bundles/main-9872389793428347239847238947.js
  output: {
    path: path.resolve('./assets/bundles/'),
    filename: '[name]-[hash].js',
  }
```

and in my django templates I use the webpack_loader template tag which looks for an emitted `webpack-stats.json` file that contains webpacks build map.

```html

<!--base.html-->
{% load render_bundle from webpack_loader %}
...
{% block javascript %}
{% render_bundle 'main' %}
{% endblock javascript %}



<!--products.html-->
{% load render_bundle from webpack_loader %}
...
{% block javascript %}
{{ block.super }}
{% render_bundle 'app-products' %}
{% endblock javascript %}



<!--users.html-->
{% load render_bundle from webpack_loader %}
...
{% block javascript %}
{{ block.super }}
{% render_bundle 'app-users' %}
{% endblock javascript %}

```
